### ISSUE TITLE
Detection using research sensors

### ISSUE THREAD

**Issue by franckies** [doughtmw/ArUcoDetectionHoloLens-Unity#2]
Hello, first of all thanks for this cool project! I tried your sample without any modification, the cube correctly appears on top of the marker. 
I'm currently trying to detect the marker using the light right front camera, without any success.
I followed the steps you included to use research mode sensors, but when i run i obtain the following exceptions:

ArUcoDetectionHoloLensUnity.ArUcoMarkerDetection:Start()
...
HoloLensForCVUnity.ArUcoDetection.StartHoloLensMediaFrameSourceGroup: Starting the media frame source group
UnityEngine.DebugLogHandler:Internal_Log(LogType, String, Object)
UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
...
Exception thrown at 0x7743C7D2 (KernelBase.dll) in Aruco_Hololens.exe: 0x40080202: WinRT transform error (parameters: 0x8000000B, 0x80070490, 0x00000014, 0x00DCD4E8).
'Aruco_Hololens.exe' (Win32): caricamento di 'C:\Windows\System32\ksuser.dll' completato. 
...
MediaFrameSourceGroup::TryInitializeMediaCaptureAsync: MediaCapture is successfully initialized in shared mode.
...
Exception thrown at 0x7743C7D2 (KernelBase.dll) in Aruco_Hololens.exe: WinRT originate error - 0xC00DABE0 : 'No capture devices are available.'.
...
MediaFrameSourceGroup::GetSensorType:: found sensor name 'Visible Light Right-Front' in MF_MT_USER_DATA (blob has 52 bytes)
MediaFrameSourceGroup::GetSubtypeForFrameReader: evaluating MediaFrameSourceKind::Depth with format Video-ARGB32 @30/1Hz
...
Exception thrown at 0x5CCAE1FE (HoloLensForCV.dll) in Aruco_Hololens.exe: 0xC0000005: Access violation reading location 0x00000020.

Could you please help me?

**Comment by ardaegeunlu** [comment]
Have you enabled the research mode, https://docs.microsoft.com/en-us/windows/mixed-reality/research-mode ?

**Comment by franckies** [comment]
Yes research mode enabled and also added to package xml manifest all needed components. I went on doing other tests and I used getLatestSensorFrame from light right front camera before launching aruco detection , and then with some checks it turned out that the sensor frame variable was null, i.e. I suppose that the frame was not acquired correctly. 

Maybe if I use the light cameras I need some other functions? I really don’t get what I’m missing.

**Comment by franckies** [comment]
This is what i get from log:

About ArucoTracker:
Exception thrown at 0x7741C7D2 in New Unity Project.exe: Microsoft C++ exception: Il2CppExceptionWrapper at memory location 0x02AFEA24.
NullReferenceException: Object reference not set to an instance of an object.
  at ArUcoDetectionHoloLensUnity.ArUcoMarkerDetection.UpdateHoloLensMediaFrameSourceGroup () [0x00000] in <00000000000000000000000000000000>:0 
  at ArUcoDetectionHoloLensUnity.ArUcoMarkerDetection.Update () [0x00000] in <00000000000000000000000000000000>:0 

using the code:

```
await _rightFrontMediaFrameSourceGroup.StartArUcoMarkerTrackerAsync(
                markerSize, 
                (int)arUcoDictionaryName, 
                _unityCoordinateSystem);

```
About latestSensorFrame:

****************latest camera frame null*******
UnityEngine.DebugLogHandler:Internal_Log(LogType, String, Object)
UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:Log(Object)
ArUcoDetectionHoloLensUnity.ArUcoMarkerDetection:UpdateHoloLensMediaFrameSourceGroup()
ArUcoDetectionHoloLensUnity.ArUcoMarkerDetection:Update()


using the code:


```
SensorFrame latestRightFrontCameraFrame =
                _rightFrontMediaFrameSourceGroup.GetLatestSensorFrame(
                _sensorTypeResearch);

            if (latestRightFrontCameraFrame == null){
                Debug.Log("****************latest camera frame null*******");
                return;}
            else{
                Debug.Log("***********************frame not null!*****************");}

```

here the full log:
[logwithlatestframe.txt](https://github.com/doughtmw/ArUcoDetectionHoloLens-Unity/files/4337525/logwithlatestframe.txt)

**Comment by franckies** [comment]
Going on with the analysis the problem seems to be here:

```
MediaFrameSourceGroup::GetSubtypeForFrameReader: evaluating MediaFrameSourceKind::Depth with format Video-ARGB32 @30/1Hz

MediaFrameSourceGroup::InitializeMediaSourceWorkerAsync: no eligible sources in 'Sensor Streaming'
```
If I try all other sensors they are correctly enabled and initialised, but all the four ambientale cameras  do have problems with InitializeMediaSourceWorkerAsync. 

Any suggestions?

**Comment by franckies** [comment]
It seems that @doughtmw forced the format Color to be used only with PV camera here:

   ```
switch (kind)
        {

        case Windows::Media::Capture::Frames::MediaFrameSourceKind::Color:

			// Force set the media stream to the desired format
			// Ensure we select the desired video stream, check width and height
			// MediaFrameSourceInfo->Source->Profile  Width : 896, Height : 504, FrameRate : 29.970030
			if (format->VideoFormat->Width == 896 && format->VideoFormat->Height == 504)
				//if (true)
			{
#ifDBG_ENABLE_INFORMATIONAL_LOGGING
				dbg::trace(
					L"MediaFrameSourceGroup::GetSubtypeForFrameReader: evaluating MediaFrameSourceKind::Color with format %s-%s @%i/%iHz and resolution %i x %i",
					format->MajorType->Data(),
					format->Subtype->Data(),
					format->FrameRate->Numerator,
					format->FrameRate->Denominator,
					format->VideoFormat->Width,
					format->VideoFormat->Height);
#endif /* 
DBG_ENABLE_INFORMATIONAL_LOGGING */

				//
				// For color sources, we accept anything and request that it be converted to Bgra8.
				//
				return Windows::Media::MediaProperties::MediaEncodingSubtypes::Bgra8;
			} 
```

Can I just remove the ‘if’ and rebuild the library?

**Comment by doughtmw** [comment]
Hi, @franckies. Thanks for your question! Yes, you can absolutely replace the if statement which forces the color and frame profile. I was using that to enforce the low-resolution profile for doing some real-time processing stuff but forgot to remove it from the public repo.

**Comment by franckies** [comment]
Hi @doughtmw , thank for you reply. Actually i'm trying to remove it and re-compile the solution but without any success.
I'm following these steps:
1. Open your personal branch of Hololens for CV https://github.com/doughtmw/HoloLensForCV
2.  Installing your Nuget Packages https://github.com/doughtmw/NuGet-Package-Creation/blob/master/OpenCV.UWP.411.0.0.nupkg, without removing others packages (I'm doing it right?, because i find duplicates in output folder when I compile the Hololens for cv project, i.e. both aruco411.dll and aruco.dll)
3. Build Hololens for CV Debug x86
I do not obtain any error, but in the output folder i've duplicate dll, and moreover if I remove the OpenCV.Hololens nuget package, leaving just the OpenCV.UWP package, in the output folder there are just aruco.dll and other files (not aruco411.dll, that is the one you included in the sample of your project).

When I build the project from unity and deploy on Hololens the sensor streamer () cannot be correctly instantiated and the program goes in loop because sensorstreamer and mediaframesourcegroup is null.

I guess i'm doing something wrong in rebuilding the library but i don't know how to fix it. 
Could you help me?

It would be enough to simply re-commit the sample with the fixed plugins containing the winmd compiled without the if statement.
Thank you in advance.

**Comment by doughtmw** [comment]
Hi @franckies! I made some changes in the most recent commit to remove the forced media frame source kind for streaming. I also modified the C# sample to allow for the selection and correct initialization of research mode sensors. The problem now is that I'm not sure how to access the camera intrinsic parameters for the research mode sensors, so you would need to add a bypass for the ArUco sample to work with anything other than the Photo-Video camera.
